src/main.rs

```diff
@@ -261,7 +261,7 @@ pub struct TopLevel {
     dbus_system_bus: bool,
 
     #[argh(option, long = "event-monitor")]
-    /// path=<path/to/a/file>|fd=<fd>
+    /// broadcast, path=<path/to/a/file>|fd=<fd>
```

Comment:
Yeah, understandably so: this option is actually useless at the moment. If the D-Bus API isn't enabled, the receiver half is dropped (since no one uses it), and calls to send on the sender part simply fail silently, not causing any problems.

I think the best way forward would be to remove this suboption since we automatically enable the `event-monitor` when the D-Bus API is enabled. Users can enable the file-based monitor optionally by using `--event-monitor path=<path>`. What do you think?

Also, do you have any input on the use of `flume` as the MPMC provider? I could modify this implementation to use an MPSC instead and broadcast events only to the D-Bus API, but I'm not sure if that's a good design choice.

Reply:
> Yeah, understandably so: this option is actually useless at the moment. If the D-Bus API isn't enabled, the receiver half is dropped (since no one uses it), and calls to send on the sender part simply fail silently, not causing any problems.
> 
> I think the best way forward would be to remove this suboption since we automatically enable the `event-monitor` when the D-Bus API is enabled. Users can enable the file-based monitor optionally by using `--event-monitor path=<path>`. What do you think?
> 

Sounds good to me

> Also, do you have any input on the use of `flume` as the MPMC provider? I could modify this implementation to use an MPSC instead and broadcast events only to the D-Bus API thread, but I'm not sure if that's a good design choice.

If only one thread at the moment needs is then i would prefer not to have the extra dependency. It can always be added if more than one thread needs it. For the file based event monitor do we currently block the VMM for writing those event messages (f so, we should create an issue to do that from its own thread or switch to `O_NONBLOCKING` as I don't think we want the VMM blocked like that. If we had that second thread - maybe then we would need to use flume.

Reply:
> For the file based event monitor do we currently block the VMM for writing those event messages?

The file mode is set to non-blocking in the [`set_monitor`](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/d760a77fbb358318a84b54cebc6182e4491ea025/event_monitor/src/lib.rs#L23) function.

But I believe the [current implementation](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/d760a77fbb358318a84b54cebc6182e4491ea025/event_monitor/src/lib.rs#L55-L58) is prone to data races since `event_monitor` is being called from multiple threads, and we don't do any synchronization - which would have been blocking in nature. Maybe we do need to give `event_monitor` its own thread. 

Here's an excerpt from the events - I injected the caller thread info into each event instance to get this (meaning this is not a full trace):

```
$ tail -F events.log | grep Thread
    "thread-info": "Thread { id: ThreadId(1), name: Some(\"main\"), .. }"
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }"
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }"
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }",
    "thread-info": "Thread { id: ThreadId(5), name: Some(\"vcpu0\"), .. }"
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }",
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }",
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }",
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }"
    "thread-info": "Thread { id: ThreadId(2), name: Some(\"vmm\"), .. }"
```

Reply:
Right. I can see that we will have data races from accessing the event_montior file concurrently. Looks like it is better to move the event monitor to its own thread using blocking access to the file? Of course, it will also require `flume`.

Reply:
Also, I believe the unsafe block comment above are wrong [1], which is the reason why the racing here was allowed and not properly addressed. The unsafe is not about  whether the `Option` is valid, instead it is about allowing mutable statics (e.g. aliasing violation).

[1] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/d760a77fbb358318a84b54cebc6182e4491ea025/event_monitor/src/lib.rs#L47

Reply:
> Looks like it is better to move the event monitor to its own thread using blocking access to the file? Of course, it will also require `flume`.

Definitely onboard with doing this. I might still be able to do it without using `flume` or any alternative dependency by using two MPSCs in the implementation: one for sending the event to the `event-monitor` thread and one for sending the event to the D-Bus API thread.

I also very recently realized that the MPMC channel exposed by `flume`, despite its name, doesn't actually broadcast a value sent by the sender to all `Receiver` instances. Instead, only a single `Receiver` consumes the value. This works well for sharing work among threads for example, but it's not useful for the scenario we have here. This wasn't a problem when I was testing it with the implementation I have here because there was only one consumer but it become clear when I actually needed to use a broadcasting channel in a recent project of mine. Terribly sorry about this.

Going forward, if everyone is onboard with it, I can start working on moving `event-monitor` into its own thread and try implementing the feature in this branch without using an external dependency. I will only include one if I find the implementation I arrive at unfit. What do you guys think?

Reply:
> Going forward, if everyone is onboard with it, I can start working on moving `event-monitor` into its own thread and try implementing the feature in this branch without using an external dependency. I will only include one if I find the implementation I arrive at unfit. What do you guys think?

Sounds good to me.

Reply:
PR is live here: #5633, I'll revise the implementation in this branch once/if that PR gets merged.